include/oneapi/dpl/experimental/kt/internal/esimd_radix_sort_kernels.h

```diff
@@ -0,0 +1,706 @@
+
+    // max SLM is 256 * 4 * 64 + 256 * 2 * 64 + 257*2, 97KB, when  __data_per_work_item = 256, __bin_count = 256
+    // to support 512 processing size, we can use all SLM as reorder buffer with cost of more barrier
+    __dpl_esimd::__ns::slm_init<::std::max(__reorder_slm_size, __bin_hist_slm_size + __incoming_offset_slm_size)>();
+
+    const ::std::uint32_t __local_tid = __idx.get_local_linear_id();
+
+    const ::std::uint32_t __slm_reorder_this_thread = __local_tid * __data_per_work_item * sizeof(_KeyT);
+    const ::std::uint32_t __slm_bin_hist_this_thread = __local_tid * __hist_stride;
+
+    __dpl_esimd::__ns::simd<_HistT, __bin_count> __bin_offset;
+    __dpl_esimd::__ns::simd<_DeviceAddrT, __data_per_work_item> __write_addr;
+    __dpl_esimd::__ns::simd<_KeyT, __data_per_work_item> __keys;
+    __dpl_esimd::__ns::simd<_BinT, __data_per_work_item> __bins;
+    __dpl_esimd::__ns::simd<_DeviceAddrT, __data_per_step> __lane_id(0, 1);
+
+    const _DeviceAddrT __io_offset = __data_per_work_item * __local_tid;
+
+    static_assert(__data_per_work_item % __data_per_step == 0);
+    static_assert(__bin_count % 128 == 0);
+    static_assert(__bin_count % 32 == 0);
+    _ONEDPL_PRAGMA_UNROLL
+    for (::std::uint32_t __s = 0; __s < __data_per_work_item; __s += __data_per_step)
+    {
+        __dpl_esimd::__ns::simd_mask<__data_per_step> __m = (__io_offset + __lane_id + __s) < __n;
+        __keys.template select<__data_per_step, 1>(__s) =
+            __dpl_esimd::__ns::merge(__dpl_esimd::__gather<_KeyT, __data_per_step>(__rng_data(__rng_pack.__keys_rng()), __lane_id, __io_offset + __s, __m),
+                  __dpl_esimd::__ns::simd<_KeyT, __data_per_step>(__sort_identity<_KeyT, __is_ascending>()), __m);
+    }
+
+    for (::std::uint32_t __stage = 0; __stage < __stage_count; __stage++)
+    {
+        __bins = __get_bucket<__mask>(__order_preserving_cast<__is_ascending>(__keys), __stage * __radix_bits);
+
+        __bin_offset = 0;
+        _ONEDPL_PRAGMA_UNROLL
+        for (::std::uint32_t __s = 0; __s < __data_per_work_item; __s += 1)
+        {
+            __write_addr[__s] = __bin_offset[__bins[__s]];
+            __bin_offset[__bins[__s]] += 1;
+        }
+
+        /*
+        first write to slm,
+        then do column scan by group, each thread to 32c*8r,
+        then last row do exclusive scan as incoming offset
+        then every thread add local sum with sum of previous group and incoming offset
+        */
+        {
+            __dpl_esimd::__ns::barrier();
+            _ONEDPL_PRAGMA_UNROLL
+            for (::std::uint32_t __s = 0; __s < __bin_count; __s += 128)
+            {
+                __dpl_esimd::__block_store_slm<::std::uint32_t, 64>(
+                    __slm_bin_hist_this_thread + __s * sizeof(_HistT),
+                    __bin_offset.template select<128, 1>(__s).template bit_cast_view<::std::uint32_t>());
+            }
```

Comment:
> Is there some restrictions on the type for __block_store_slm? Why cant we just store 128 elements of type _HistT?

There used to be a restriction on the types. Intel® oneAPI DPC++/C++ Compiler Release Notes (2023.1):
_Added 8/16-bit type support to lsc_block_load and lsc_block_store ESIMD API._

ESIMD sort KT is going to supported for 2023.2 and newer, so we can handle `_HistT` directly.

> If this isn't possible, we should calculate the size difference between _HistT and ::std::uint32_t and use that rather than assuming a factor of 2.

Agree. It will save us from issues when changing the code in the future.